sdk/src/main/scala/com/horizen/storage/SidechainStateStorage.scala

```diff
@@ -171,12 +194,24 @@ class SidechainStateStorage(storage: Storage, sidechainBoxesCompanion: Sidechain
     updateList.add(new JPair(withdrawalEpochInformationKey,
       new ByteArrayWrapper(WithdrawalEpochInfoSerializer.toBytes(withdrawalEpochInfo))))
 
+    val isWithdrawalEpochSwitched: Boolean = getWithdrawalEpochInfo match {
+      case Some(storedEpochInfo) => storedEpochInfo.epoch != withdrawalEpochInfo.epoch
+      case _ => false
+    }
     if (withdrawalRequestAppendSeq.nonEmpty) {
-      val withdrawalRequestSeq = getWithdrawalRequests(withdrawalEpochInfo.epoch) ++ withdrawalRequestAppendSeq
+      // Calculate the next counter for storing withdrawal requests without duplication previously stored ones.
+      val nextWithdrawalEpochCounter: Int = if(isWithdrawalEpochSwitched) 0 else getWithdrawalEpochCounter.getOrElse(-1) + 1
+
+      updateList.add(new JPair(getWithdrawalEpochCounterKey, new ByteArrayWrapper(Ints.toByteArray(nextWithdrawalEpochCounter))))
 
-      updateList.add(new JPair(getWithdrawalRequestsKey(withdrawalEpochInfo.epoch),
-        new ByteArrayWrapper(withdrawalRequestSerializer.toBytes(withdrawalRequestSeq.asJava))))
+      updateList.add(new JPair(getWithdrawalRequestsKey(withdrawalEpochInfo.epoch, nextWithdrawalEpochCounter),
+        new ByteArrayWrapper(withdrawalRequestSerializer.toBytes(withdrawalRequestAppendSeq.asJava))))
 
+      // If withdrawal epoch switched to the next one, then remove outdated withdrawal related records (2 epochs before).
+      if (isWithdrawalEpochSwitched) {
+        for (i <- 0 to getWithdrawalEpochCounter.getOrElse(-1))
```

Comment:
we should use the specific counter for that epoch and then remove also the counter